### PR TITLE
Generational stack scanning

### LINKED ISSUE #13574: Restore already scanned optimisation
OCaml 4 has "Already Scanned" optimisation to avoid rescanning stack frames that haven't changed. See https://github.com/ocaml/ocaml/blob/9eb434c0eb56895add0e3a3e1fd3b3afe33eaf0a/runtime/roots_nat.c#L312. This feature was not implemented in OCaml 5 only because initially OCaml 5 only ran on x86, where this feature was not available. There are no blockers to this feature being re-enabled.

In OCaml 5, since the runtime maintains the stacks, the stack limits are generous: https://github.com/ocaml/ocaml/pull/11238. Hence, the already scanned optimisation is much more useful than in OCaml 5 than in OCaml 4. So much so that it may be worth coming with a scheme to enable this feature in all the architectures and not just in those where bits can be stolen return addresses. One possibility is to have an extra word in every frame for the already scanned information. Since only a single bit is necessary for the already scanned information, the rest of the bits may be used for other purposes.

CC @stedolan and @jasigal, who has some OCaml programs that may benefit from this optimisation.

### PR DESCRIPTION
**Pull request by xavierleroy**
Following discussions at #11923 and the request at #13574, this PR PR is a first step towards generational scanning of stack frames during minor GCs: scanning can stop when it reaches a frame that has not changed since the previous minor GC.

The first commit 97ac9a5012 simply ports the OCaml 4 approach to OCaml 5.  Already-scanned frames are identified by an otherwise-ignored bit in the corresponding return addresses stored in the stack.  This is supported for POWER, RISC-V, and ARM 64-bit in top-bits-ignore mode, that is, currently, ARM64/Linux but not ARM64/macOS.  The nice thing about this approach is that it incurs zero performance overhead on the mutators, and requires no extra stack space.

Commit  e8f7ea1b46 refactors the OCaml 4 code a bit to make it more applicable to other marking scheme later.  It also opens the possibility to detect ignored bits in the configure script rather than hard-wiring their existence in `caml/stack.h`.

Commit d9145d98c6 is a failed experiment to implement top-bits-ignore for x86-64 by catching SIGSEGV arising out of RET instructions and fixing them in the signal handler.  It is prohibitively slow, so 62a9b88ab0 reverts this commit.  I left the code temporarily in this PR as a warning sign for those of you who would be tempted to do things with signals here.

Commit c960af4753 implements a form of top-bits-ignore for ARM64/macOS, just by masking the top 8 bits of the return address before every `ret` instruction.  Unlike hardware TBI, it does not come for free: code size increases by 2% (for the OCaml stdlib) and execution time by 0.5% on an average of small benchmarks.  The +0.5% does not include the potential reduction in GC time offered by generational stack scanning.  Maybe Sandmark can give better performance estimates.

~~I trieds a similar "mask bits off before returning" approach for x86-64 but the performance figures were all over the place, so I probably did something wrong.~~

Marking this PR as a draft because there are other approaches to explore, this is just to get the ball rolling.

Closes: #13574.

### CODE REVIEW

**Comment by xavierleroy** [comment]
bd3b7b4 implements "bit 60 ignore" before x86-64 RET instructions.  The impact is more negative than for ARM 64: code size increases by 3%, but execution time by 7.5% on an average, with several benchmarks increasing by more than 20%.  This doesn't work, so 5f5c65d reverts this code.

**Comment by stedolan** [comment]
For x86-64, here is a somewhat disgusting scheme if you're interested:

  - Ensure with assembler directives that the _end_ of every call instruction lands on an even address. (Requires knowing the lengths of the various kinds of call instructions, but at least you can make the assembler verify that you're right with the trick in #11162)
  - Put a one-byte NOP instruction after every call
  - Use the _low_ bit of the return address to indicate "already scanned": either value will work fine, as control will land either before or after the NOP.

Changing the actual return addresses like this will throw off the return address predictor, hurting performance. So if you do have a go at this, I recommend only applying the mark to stack frames at least, say, 16 frames above the current stack pointer. (The generational stack scanning optimisation is not useful for very recent calls, and the return address predictor overflows quickly so is only useful for very recent calls).

**Comment by kayceesrk** [comment]
The return address low bit mangling on x86 may interfere with Intel CET shadow stack [1] 

> This helps ensure that even when an attacker is able to modify/corrupt the return addresses in the data stack for the purpose of carrying out a ROP attack, the attacker is not able to modify the Shadow Stack, and the CET state machine in the CPU detects mismatches between the address on the shadow and data stack to help prevent the attack via an exception reported to the OS.

[1] https://www.intel.com/content/www/us/en/developer/articles/technical/technical-look-control-flow-enforcement-technology.html

**Comment by xavierleroy** [comment]
> The return address low bit mangling on x86 may interfere with Intel CET shadow stack

That was my reaction too.  This said, the way OCaml compiles exceptions and effects is not obviously compatible with a hardware shadow stack either.

**Comment by kayceesrk** [comment]
> This said, the way OCaml compiles exceptions and effects is not obviously compatible with a hardware shadow stack either.

Yes, indeed, you are right. We'd be no worse than we are today.

**Comment by NickBarnes** [comment]
Do we support shadow-stacks? How do they work with fiber stacks?

**Comment by xavierleroy** [comment]
No, we don't support shadow stacks at the moment, and it would take some work to accommodate exceptions and effects ("fibers").  But maybe we should think twice before making OCaml even more incompatible with shadow stacks than it is today.

**Comment by chambart** [comment]
I have been playing a bit with @stedolan's version (in https://github.com/chambart/ocaml-1/tree/generational-stack-scanning)
The code size increase is noticeable (+1.5 nops per call), this seems to be ~10% for small benchmarks. for `ocamlc.opt` this is around 20%.
I tested some small benchmarks which show +-5% of differences. `ocamlc.opt` is ~3% faster.

I'm not sure that the cost of messing with the return address prediction will be so large. Most of the calls are still going thought the normal return address. If we want to reduce the potential cost returns that are likely to be used soon, we might want to mark only the return addresses after a few stack frames.
But the increase in code size can be quite costly.

**Comment by gasche** [comment]
My broad opinion (expressed offline but not online, it seems) is that we should merge this support for the architectures for which it comes at no cost, as a first step. I would welcome a first PR that focuses on those (basically it corresponds to the current diff, with a simplified history) to review for correctness and merge, and then we can keep discussing clever schemes for x86-64.

**Comment by xavierleroy** [comment]
> My broad opinion (expressed offline but not online, it seems) is that we should merge this support for the architectures for which it comes at no cost, as a first step

I'm fine with that.  But I'd still like to get detailed performance numbers for the ARM64/macOS explicit untagging of return addresses.  Does anyone know how to run Sandmark on a Mac?

**Comment by OlivierNicole** [comment]
From a distance the tools and libraries it depends on do not _seem_ Linux-specific so I assume one can clone the repo and follow the instructions (except for the dependencies part) to run the benchmarks locally. I am not entirely sure, however. I have asked around at Tarides in case someone knows better. (Maybe @punchagan?)

**Comment by kayceesrk** [comment]
> I'd still like to get detailed performance numbers for the ARM64/macOS explicit untagging of return addresses. Does anyone know how to run Sandmark on a Mac?

I've run Sandmark on a Linux ARM machine: https://github.com/ocaml/ocaml/pull/13393#issuecomment-2311660507. I had to remove the owl benchmarks (which are only available on x86) to get them to work (by deleting the lines here: https://github.com/ocaml-bench/sandmark/blob/main/dependencies/template/dev-5.x.opam#L37-L38). I suspect no one has given macOS runs a go.

**Comment by tmcgilchrist** [comment]
🤚🏻 I have run the tests on macOS earlier this year, I can't recall how many changes I needed to make. I will check today. However the owl tests should be working now on macOS/arm64, check that you get `owl.1.1`.  https://github.com/owlbarn/owl/pull/606

**Comment by xavierleroy** [comment]
I tried to run Sandmark on a Mac M1 and gave up after 30 minutes.
- The instructions ask to install C dependencies using `apt-get` and Debian package names.  I found Homebrew equivalents for the more obvious packages, and gave up on others.
- Then, `bash run_all_serial.sh` complains about a missing file for 5.1.0+trunk.  Replacing by 5.4.0+trunk.
- Then, the script re-checks C dependencies using `dkpg` and reports many errors (of course!)
- Then, it tries to use `grep` with the `-P` option, which is a GNU-ism.
- Then, it aborts on a `case in *5*)` that looks like an unprotected empty shell variable.

Should we conclude that Sandmark can only run under Debian/Ubuntu?

**Comment by xavierleroy** [comment]
I cleaned up this PR and I think it is now ready for review.

I ran additional tests on a Mac M1 using Coq (as an example of an OCaml program that does a lot of GCs but avoids deep recursion (AFAIK)).  Code size increases by 1.3% owing to the explicit bit masking before `ret` instructions, but the changes in execution times are lost in the noise.  Artificial tests that recurse very deep run significantly faster, as expected.

**Comment by ghost** [comment]
The use of `Saved_return_address` in `runtime/tsan.c` also needs to be adjusted:
```
diff --git runtime/tsan.c runtime/tsan.c
index 428c716ad..801215458 100644
--- runtime/tsan.c
+++ runtime/tsan.c
@@ -340,10 +340,13 @@ CAMLno_tsan void caml_tsan_entry_on_resume(uintnat pc, char* sp,
       return;
     }
 
     sp = First_frame(stack->sp);
     next_pc = Saved_return_address(sp);
+#ifdef Mask_already_scanned
+    next_pc = Mask_already_scanned(next_pc);
+#endif
   }
 
   caml_tsan_entry_on_resume(next_pc, sp, stack);
   caml_tsan_debug_log_pc("forced__tsan_func_entry for", pc);
   __tsan_func_entry((void*)next_pc);
```

However, even with this change, there are two tests which still fail on arm with tsan-enabled build: `tests/effect-syntax/coroutines.ml` and `tests/effects/unhandled_effects.ml`, both cause a stack overflow in `caml_tsan_entry_on_resume`. Currently investigating.

**Comment by ghost** [comment]
The complete fix is
```
diff --git runtime/tsan.c runtime/tsan.c
index 428c716ad..1a57fc527 100644
--- runtime/tsan.c
+++ runtime/tsan.c
@@ -330,20 +330,26 @@ void caml_tsan_exit_on_perform(uintnat pc, char* sp)
 CAMLno_tsan void caml_tsan_entry_on_resume(uintnat pc, char* sp,
     struct stack_info const* stack)
 {
   caml_frame_descrs* fds = caml_get_frame_descrs();
   uintnat next_pc = pc;
+#ifdef Mask_already_scanned
+  next_pc = Mask_already_scanned(next_pc);
+#endif
 
   caml_next_frame_descriptor(fds, &next_pc, &sp, (struct stack_info*)stack);
   if (next_pc == 0) {
     stack = stack->handler->parent;
     if (!stack) {
       return;
     }
 
     sp = First_frame(stack->sp);
     next_pc = Saved_return_address(sp);
+#ifdef Mask_already_scanned
+    next_pc = Mask_already_scanned(next_pc);
+#endif
   }
 
   caml_tsan_entry_on_resume(next_pc, sp, stack);
   caml_tsan_debug_log_pc("forced__tsan_func_entry for", pc);
   __tsan_func_entry((void*)next_pc);
```
and with this version, there are no longer test regressions when using tsan.

**Comment by xavierleroy** [comment]
Thanks @dustanddreams for the work on TSAN !  It made me realize that this PR can be made less intrusive by masking the mark off return addresses in `Saved_return_address` (commit 062b1cf3dc).  However your second fix is still needed, I put it in commit 69b90a842c.  I tested on macOS with TSAN enabled, and got a bunch of errors, but I'm not sure TSAN is supported there.  I'm currently testing on ARM64/Linux with TSAN enabled.

**Comment by ghost** [comment]
Your changes to wrap the return address macro make sense and indeed help reading/reviewing. There should be no regressions in tsan builds in your latest version of this PR.

**Review by ghost — APPROVED** [review]
Looking good at this point, and appears to work as intended.

**Comment by xavierleroy** [comment]
Rebased on current trunk and ready to merge in 5.4.  A quick yes/no decision would be appreciated.

File: asmcomp/arm64/emit.mlp [hunk 1]
```diff
@@ -983,7 +983,12 @@ let emit_instr env i =
     | Lreloadretaddr ->
         ()
     | Lreturn ->
-        output_epilogue env (fun () -> `	ret\n`)
+        output_epilogue env begin fun () ->
+          if not top_bits_ignore then
+            (* mask the top 8 bits ourselves *)
+            `	and	x30, x30, #0x00FFFFFFFFFFFFFF\n`;
+          `	ret\n`
```
Comment (gasche) [inline asmcomp/arm64/emit.mlp:991]: Nitpick: on first read I was convinced that you forgot the `else` case and that no `ret` would be emitted in the `top_bit_ignore` case. I would be in favor of a `then begin ... end` here for readability.

File: runtime/tsan.c [hunk 2]
```diff
@@ -332,6 +332,9 @@ CAMLno_tsan void caml_tsan_entry_on_resume(uintnat pc, char* sp,
 {
   caml_frame_descrs* fds = caml_get_frame_descrs();
   uintnat next_pc = pc;
+#ifdef Mask_already_scanned
+  next_pc = Mask_already_scanned(next_pc);
+#endif
```
Comment (gasche) [inline runtime/tsan.c:356]: I have a bad gut feeling about this change.

1. There are several other `caml_tsan_exit_on_*` functions that look very similar to my untrained eye, and I don't understand why they don't need the same change.
2. The change suggests that the input argument should be typed liked a "raw" pc address, but the function calls itself recursively with a "masked" pc address. I wish it would be clear which of the type is expected -- and ideally the type discipline would be the same for all `tsan_{entry,exit}_*` functions in the module.

Additional questions about this part of the code that I don't understand very well:
- does `caml_next_frame_descriptor` expect a masked pc or a raw pc in input? (it writes a masked pc back)
- `caml_stash_backtrace` is called from the assembly runtime files in a way very similar to `caml_tsan_entry_on_resume`, does it need a new masking instruction as well?
- I wonder if `__tsan_func_entry`, which is called in the assembly runtime files by `TSAN_ENTER_FUNCTION` macros, also needs the masking

**Review by gasche — COMMENTED** [review]
I have tried to review this to help along. This looks generally fine, but I have naive questions about what is the specification when the runtime routines in assembly call C functions which expect the return address as argument (mostly TSAN functions, but also `caml_stash_backtrace` if I understand correclty).

File: runtime/tsan.c (hunk 2 above)
Comment (xavierleroy) [reply runtime/tsan.c:356]: I cannot help with understanding the TSAN code.  But to answer your other questions:

> does caml_next_frame_descriptor expect a masked pc or a raw pc in input? (it writes a masked pc back)

It expects an unmarked PC.  (See below for the distinction between "masked" and "unmarked".   Freshly-generated return addresses are unmarked and don't need masking.)

> caml_stash_backtrace is called from the assembly runtime files in a way very similar to caml_tsan_entry_on_resume, does it need a new masking instruction as well?

No, because marked return address can only come from the OCaml stack (after they have survived a minor GC).  "Fresh" return addresses generated by a "call" instruction are always unmarked.  The PC given as argument to  `caml_stash_backtrace` by `caml_raise_exn` et al is a fresh return address.

Comment (gasche) [reply runtime/tsan.c:356]: Thanks for the explanations. I still don't understand why `caml_tsan_entry_on_resume` does explicit unmasking, but the other tsan functions don't. I currently have two hypotheses:

1. Maybe the code is wrong because `caml_tsan_exit_on_{raise,perform}` could also be passed a marked pc, so they should also be protected by explicit unmasking.
2. Maybe the reason why @dustanddreams added explicit masking on the entry of `caml_tsan_entry_on_resume` is *not* that it could be called from the assembly runtime with a marked pc, but that (with the original version you wrote) it could be called *recursively* with a marked pc, in the case where `caml_next_frame_descriptor` returned a non-zero pc. If that was the reason, then the change is not needed anymore, because now `caml_next_frame_descriptor` always masks the returned pc.

My current bet is on (2), which suggests removing this change in tsan.c. @dustanddreams may be able to confirm.

Comment (gasche) [reply runtime/tsan.c:356]: (I edited my comment as it used "masked" for "marked", which has the opposite meaning.)

Comment (OlivierNicole) [reply runtime/tsan.c:356]: I am also taking a look at the TSan-related part to try and answer @gasche's questions.

**Comment by OlivierNicole** [comment]
I ran a TSan CI, before realizing that it was not relevant, since that CI runs on amd64, which should not be affected by the change. The three test failures there are unrelated to this PR and should be ignored.

File: runtime/tsan.c (hunk 2 above)
Comment (OlivierNicole) [reply runtime/tsan.c:356]: There are two properties to enforce:

a. `caml_next_frame_descriptor` should not be passed a marked pc
b. neither should `__tsan_func_entry`.

The current version with @dustanddreams's fix enforces (b).

Regarding (a), I am not entirely sure that it is true, and if it is, the demonstration relies on non-trivial arguments and complicated invariants of how and where each of these TSan-related functions can be called. (I have painful memories of hours of debugging with @fabbing.) Such invariants would easily break with a small change in the runtime.

For this reason I would recommend to untag every pc passed to `caml_next_frame_descriptor` in `tsan.c`. It enforces (a) in a robust way, and the TSan-enabled runtime is very slow anyway.

```diff
diff --git a/runtime/tsan.c b/runtime/tsan.c
index 5f75df266a..3d3e1dceed 100644
--- a/runtime/tsan.c
+++ b/runtime/tsan.c
@@ -217,6 +217,9 @@ void caml_tsan_exit_on_raise(uintnat pc, char* sp, char* trapsp)
   caml_domain_state* domain_state = Caml_state;
   caml_frame_descrs* fds = caml_get_frame_descrs();
   uintnat next_pc = pc;
+#ifdef Mask_already_scanned
+  next_pc = Mask_already_scanned(next_pc);
+#endif
 
   /* iterate on each frame  */
   while (1) {
@@ -301,6 +304,9 @@ void caml_tsan_exit_on_perform(uintnat pc, char* sp)
   struct stack_info* stack = Caml_state->current_stack;
   caml_frame_descrs* fds = caml_get_frame_descrs();
   uintnat next_pc = pc;
+#ifdef Mask_already_scanned
+  next_pc = Mask_already_scanned(next_pc);
+#endif
 
   /* iterate on each frame  */
   while (1) {
```

**Comment by xavierleroy** [comment]
> I ran a TSan CI, before realizing that it was not relevant, since that CI runs on amd64, which should not be affected by the change. 

Does OCaml + TSan work on anything other than x86-64/Linux?  What would it take to test it on ARM64 platforms?  There are ARM64/macOS runners in Github, and ARM64/Linux in the Jenkins CI.

File: runtime/tsan.c (hunk 2 above)
Comment (gasche) [reply runtime/tsan.c:356]: I don't understand why the current version would enforce that `__tsan_func_entry` gets an unmarked pc. If I understand correctly, the two ways into which `next_pc` can flow into `__tsan_func_entry` within the function `caml_tsan_entry_on_resume` are:

1. `caml_next_frame_descriptor(fds, &next_pc, &sp, (struct stack_info*)stack);`
2. `next_pc = Saved_return_address(sp);`

With the current PR (2) always returns an unmarked PC, and my understanding is that (1) also does (`caml_next_frame_descriptor` expects a possibly-marked pc and writes an unmarked pc). This suggests that we don't need any marking in this function.

Besides, there are several other calls to `__tsan_func_entry` directly from the assembly runtime, which are not protected. Given @xavierleroy's explanation I believe that they fall under the category of "addresses of functions we just called", which cannot have been marked since they were pushed on the stack. (I wondered if tail-calls could invalidate this assumption -- a situation where a tail-call would reuse the (already-marked) return address of its caller -- but in fact I think that we don't tail-call into these runtime functions?)

**Comment by gasche** [comment]
Yes, TSan works on arm, this is what @dustanddreams used in his testing reported in https://github.com/ocaml/ocaml/pull/13594#issuecomment-2505634624 . If you have an osx arm machine, it may be enough to just configure the compiler with tsan (`./configure --enable-tsan`) and then run the testsuite.

**Comment by xavierleroy** [comment]
>  If you have an osx arm machine, it may be enough to just configure the compiler with tsan (./configure --enable-tsan) and then run the testsuite.

I tried earlier and it didn't pass the test suite.  I just tried again and it doesn't finish the build: the generated `ocamlc.opt` segfaults...  No idea what is going on.

(EDIT: the crash is with the OCaml trunk; I didn't try this PR.)

**Comment by xavierleroy** [comment]
But, again, there are ARM-macOS runners in Github, so the tsan.yml action could be extended to run there in addition to on x86.

File: runtime/tsan.c (hunk 2 above)
Comment (OlivierNicole) [reply runtime/tsan.c:356]: OK, the change in `caml_tsan_entry_on_resume` may be unnecessary to unmark the argument of `__tsan_func_entry`, however it is necessary to avoid passing a marked pc to `caml_next_frame_descriptor`.

> Besides, there are several other calls to __tsan_func_entry directly from the assembly runtime, which are not protected. Given @xavierleroy's explanation I believe that they fall under the category of "addresses of functions we just called", which cannot have been marked since they were pushed on the stack.

I have looked at `arm64.S` and it seems ok to me, all the calls are from routines where there can be no minor GC between the push of the return address and the call to `__tan_func_entry`.
However, I am not competent regarding POWER and RISC-V.

> (I wondered if tail-calls could invalidate this assumption -- a situation where a tail-call would reuse the (already-marked) return address of its caller -- but in fact I think that we don't tail-call into these runtime functions?)

Yes, tail calls do require special attention, and sometimes we do tail call these routines—fortunately, in `arm64.S`, not from an OCaml stack.

**Comment by OlivierNicole** [comment]
Yes, you should be able to replace `ubuntu-latest` by `macos-latest` in `tsan.yml` in this PR. It should also install libunwind.

**Comment by OlivierNicole** [comment]
If this is one day extended to amd64, `amd64.S` will require a review, for instance this is an unprotected `__tsan_func_entry` that may be passed a marked address.
https://github.com/ocaml/ocaml/blob/ce310157f0bb16875cf925aebdac71f0a4b0f967/runtime/amd64.S#L1159

File: runtime/tsan.c (hunk 2 above)
Comment (gasche) [reply runtime/tsan.c:356]: I think we should try to:

- in the short term, converge to a recommendation for @xavierleroy on the tsan part of his PR. Is it fine as-is? Should the one-line change be removed? Should all entry points start with pc masking? I would be in team "remove the one-line change" if there can be testing to confirm that it now works correctly, otherwise "protect all entry points". Ideally we could converge to a final decision and merge in the next few days.

- In the medium term, decide a discipline for the interaction between TSan and return address marking. I'm not fond of the approach where we do subtle reasoning about each code generator, without introducing clear comments that explain what is going on. For example, maybe we could stop calling TSan runtime functions directly in the assembly files and systematically go through a `caml_tsan_func_entry` wrapper in tsan.c, where we could make uniform decisions about this without tweaking each assembly runtime file. (This may also help implementing the runtime knob to turn TSan off?) This could be done in a separate PR to avoid stepping on Xavier's toes.

Comment (OlivierNicole) [reply runtime/tsan.c:356]: > * in the short term, converge to a recommendation for @xavierleroy on the tsan part of his PR. Is it fine as-is? Should the one-line change be removed? Should all entry points start with pc masking? I would be in team "remove the one-line change" if there can be testing to confirm that it now works correctly, otherwise "protect all entry points". Ideally we could converge to a final decision and merge in the next few days.

I think I have expressed what I think should be done in my comments. I’m not sure what one-line change you are referring to. It cannot be @dustanddreams's fix, without it the code is incorrect—and fails in practice, as he has reported.

**Comment by OlivierNicole** [comment]
For the record, I am not fond of this PR being pushed through at the very last moment before a release, given the non-trivial review that it requires.

> In the medium term, decide a discipline for the interaction between TSan and return address marking. I'm not fond of the approach where we do subtle reasoning about each code generator, without introducing clear comments that explain what is going on. For example, maybe we could stop calling TSan runtime functions directly in the assembly files and systematically go through a caml_tsan_func_entry wrapper in tsan.c, where we could make uniform decisions about this without tweaking each assembly runtime file. (This may also help implementing the runtime knob to turn TSan off?) This could be done in a separate PR to avoid stepping on Xavier's toes.

I think it could make sense to have a unique macro (in a separate PR as you suggest) that does the unmasking, if benchmarks show that it does not degrade performance too much (that is, in the TSan world where performance is very bad already). I am wary of the additional cost that comes with a call to a C function.

File: runtime/tsan.c (hunk 2 above)
Comment (gasche) [reply runtime/tsan.c:356]: More details:
- The one-line change I'm referring to is the diff in lines 335-337 of tsan.c above.
- The test that @dustanddreams did was before Xavier introduced masking in `Saved_return_address`. My best guess (but then I may have missed something) is that the change is now unnecessary and that a test without the change on arm+tsan would now pass.

Comment (OlivierNicole) [reply runtime/tsan.c:356]: Looking at the call sites of `caml_tsan_entry_on_resume` can be called from, I don’t think that it’s the case: for instance, it can be called from `caml_resume` with the return address stored in a fiber. That fiber may well have been scanned and the return address marked. In which case, we would pass a marked pc to `caml_next_frame_descriptor`, resulting in the frame descriptor not being found and `caml_next_frame_descriptor` returning NULL without writing a new pc. The end result is an infinite recursion, probably the cause of the TSan shadow stack overflow that @dustanddreams observed.

This infinite recursion is bad error handling and we should fix it to emit a fatal error instead, but in any case we need to pass the right pc to `caml_next_frame_descriptor` in order to keep TSan’s shadow stack correct.

Comment (gasche) [reply runtime/tsan.c:356]: Thanks for the extra explanations. As a non-assembly person, I had missed the difference between (in amd64 terms) `STACK_RETADDR(%rsp)` and `STACK_RETADDR(Stack_sp(%r10))`, which is key here.

Based on our discussion I decided to edit @xavierleroy's PR with a new commit that pushes unmasking in all `caml_tsan_*` entry points. My willpower faltered at the last minute and I ended up documenting the current state (now that I understand how to check the invariant), with assertion comments. Does that sound okay to you, or do you believe that we should really unmask in all entry points?

Comment (gasche) [reply runtime/tsan.c:356]: (I'm talking about the new commit that I just pushed, currently named b3daac45f417533bdd8acefada7cbe915e187329 )

**Comment by Octachron** [comment]
For the record, I am not sold on the idea that the feature should be in 5.4. In fact, taking in account the amount of review discussion started after the freeze, .I would rather have the change on trunk only.

File: runtime/tsan.c [hunk 3]
```diff
@@ -218,6 +218,13 @@ void caml_tsan_exit_on_raise(uintnat pc, char* sp, char* trapsp)
   caml_frame_descrs* fds = caml_get_frame_descrs();
   uintnat next_pc = pc;
 
+#ifdef Mask_already_scanned
+  // The [pc] argument comes directly from the current
+  // [caml_raise_exn] or [caml_tsan_exit_on_raise_asm] frame, so it is
+  // unmarked.
+  CAMLassert(!Already_scanned(next_pc);
```
Comment (OlivierNicole) [inline runtime/tsan.c]: It can also come from `caml_raise_exception`, a routine called when raising from C. In that case the address is read from the top of the OCaml stack and I don’t see a clear reason why it couldn’t be marked.

File: runtime/tsan.c [hunk 4]
```diff
@@ -302,6 +309,12 @@ void caml_tsan_exit_on_perform(uintnat pc, char* sp)
   caml_frame_descrs* fds = caml_get_frame_descrs();
   uintnat next_pc = pc;
 
+#ifdef Mask_already_scanned
+  // The [pc] argument comes directly from the current
+  // [caml_(re)perform] frame, so it is unmarked.
+  CAMLassert(!Already_scanned(next_pc);
+#endif
+
```
Comment (OlivierNicole) [inline runtime/tsan.c]: I’m afraid I also disagree here: if `Stdlib.Effect.perform` is called in tail position, so will be `caml_perform`, in which case `next_pc` here is the return address from the enclosing frame, with no guarantees on it—for instance, if some function `f` allocates and then tail-calls `perform`, the return address may be marked.

**Review by OlivierNicole — CHANGES_REQUESTED** [review]
I am not opposed to using asserts in principle, but here I think there are scenarios where they do fail, see my comments.

File: runtime/tsan.c (hunk 2 above)
Comment (OlivierNicole) [reply runtime/tsan.c:356]: I’m not against it in principle but I think there are incorrect cases; see my review below for the details.

Comment (gasche) [reply runtime/tsan.c:356]: Thanks! I'll update my commit with marking and more accurate documentation comments.

File: runtime/tsan.c (hunk 3 above)
Comment (gasche) [reply runtime/tsan.c]: Thanks! I documented this in the new version of my commit.

File: runtime/tsan.c (hunk 4 above)
Comment (gasche) [reply runtime/tsan.c]: I documented this as well, thanks!

**Comment by gasche** [comment]
@OlivierNicole I have taken your excellent points into account and rebased my commit. If it's okay to you, I'm happy to approve the overall PR.

**Comment by OlivierNicole** [comment]
Thanks! It looks good to me for ARM64, but I am not competent on POWER and RISC-V. That being said, there is also @dustanddreams's review.
Also, I think the worst that could happen in `riscv.S` and `power.S` is that a marked pc gets passed to `__tsan_func_entry`, which is a bug but not critical (so long as the entry/exits cancel out, the worst that can happen is a wrong line in a TSan stack trace).

**Review by gasche — APPROVED** [review]
I reviewed the PR, except for the tsan part which I co-authored and was reviewed by @OlivierNicole.

**Comment by xavierleroy** [comment]
FYI, I set up a Jenkins CI job called "precheck-tsan" that tests TSAN-enabled OCaml on 4 CI architectures that support it: x86-64/Linux, ARM64/Linux, ARM64/macOS, POWER/Linux.  It can be run manually on any branch, like "precheck".

Here are the results for the current trunk : https://ci.inria.fr/ocaml/job/precheck-tsan/3/
and the results for this PR: https://ci.inria.fr/ocaml/job/precheck-tsan/2/

Bottom line: the two builds fail the test suite on all four platforms, but this PR doesn't seem to make things worse.

**Comment by OlivierNicole** [comment]
Thanks for setting this up!

The failures on x86 Linux and ARM64 macOS are flaky TSan tests. (I’m adding to my todo list to finally fix them.) On these platforms, indeed this PR doesn’t change anything.

PowerPC Linux seems very broken in both cases, with errors that are (seemingly) completely unrelated to TSan. I’ll have to investigate this… but this PR doesn’t make it worse as you said.

Finally, I don’t know how to interpret the Java EOFException for ARM64 Linux with this PR.
(And on trunk two tests unusually timeout, which also warrants investigation.)

All in all, rather a confidence boost for this PR.

**Comment by ghost** [comment]
> PowerPC Linux seems very broken in both cases, with errors that are (seemingly) completely unrelated to TSan. I’ll have to investigate this… but this PR doesn’t make it worse as you said.

Most of these failures are caused by the sanitizer's wrappers for some of the file I/O routines being broken on PowerPC. This got eventually fixed but you'd need a recent enough toolchain to make these tests pass, and I'm not sure even the latest Linux distributions on PowerPC provide them.

File: runtime/tsan.c (hunk 2 above)
Comment (ghost) [reply runtime/tsan.c:356]: >     2. Maybe the reason why @dustanddreams added explicit masking on the entry of `caml_tsan_entry_on_resume` is _not_ that it could be called from the assembly runtime with a marked pc, but that (with the original version you wrote) it could be called _recursively_ with a marked pc, in the case where `caml_next_frame_descriptor` returned a non-zero pc. If that was the reason, then the change is not needed anymore, because now `caml_next_frame_descriptor` always masks the returned pc.

Actually, both. The first change is intended to allow it to be invoked from the runtime assembly bits; the second one, to allow recursive invocation.

**Comment by ghost** [comment]
You might want to update the reviewers list in `Changes`.
